### ISSUE TITLE
Tests are leaking state between blocks.

### ISSUE THREAD

**Issue by GeeWee** [stripe/stripe-apps#342]
**Describe the bug**
If components invoke asynchronous effects such as `useEffect`, it is very easy to get non-deterministic tests where the components leak state.

**To Reproduce**
This code with two completely identical tests will have the first test pass, but the second test always fails due to a `useEffect`
```ts
export const MyComponent = () => {
    // useState starts with 1
    const [val, setval] = useState(1);

    // And after useEffect it is 2
    useEffect(() => {
        setval(2);
        }
    )

    return <div>
        {val}
    </div>
}


describe('Reproduction', () => {
    it('identical 1', async () => {
        const thing = render(<MyComponent/>);
        const {wrapper} = thing;
        expect(wrapper).toContainText("1")
    });

    it('identical 2', async () => {
        const thing = render(<MyComponent/>);
        const {wrapper} = thing;
        expect(wrapper).toContainText("1")
    });
});
```

Yarn output:
```
 FAIL  src/views/ProductDetail.test.tsx
  Reproduction
    ✓ identical 1 (14 ms)
    ✕ identical 2 (5 ms)

  ● Reproduction › identical 2

    expect(<[RemoteRoot] />).not.toContainText(expected)

    Expected the React element:
      <[RemoteRoot] />
    With text content:
      "2"
    To contain string:
      "1"

      90 |         const thing = render(<MyComponent/>);
      91 |         const {wrapper} = thing;
    > 92 |         expect(wrapper).toContainText("1")
         |                         ^
      93 |     });
      94 | });
      95 |

      at src/views/ProductDetail.test.tsx:92:25
      at src/views/ProductDetail.test.tsx:8:71
      at Object.<anonymous>.__awaiter (src/views/ProductDetail.test.tsx:4:12)
      at Object.<anonymous> (src/views/ProductDetail.test.tsx:89:34)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
Snapshots:   0 total
Time:        2.956 s, estimated 4 s
```

**Expected behavior**
I expect that the `useEffect` or any other sort of asynchronous behaviour from one test will not affect other tests.
I expect that each component is cleaned up between tests, and that `render` will always give me a fresh component.

Failing that I expect this to be documented very clearly, as it's pretty non-obvious behaviour.

**Workarounds** 
It seems that if I flush the event queue manually after each test, the useEffect will work on the first component and I will be given a fresh component for the second test.

If I change the test code to be the following, both tests pass as expected

```ts
describe('Reproduction', () => {
    it('identical 1', async () => {
        const thing = render(<MyComponent/>);
        const {wrapper, update} = thing;
        expect(wrapper).toContainText("1")
        await update(); // - NEW LINE
        // await new Promise(resolve => setTimeout(resolve, 1000)) - Also works
    });

    it('identical 2', async () => {
        const thing = render(<MyComponent/>);
        const {wrapper} = thing;
        expect(wrapper).toContainText("1")
    });
});
```

**Desktop (please complete the following information):**: 
 - OS: [e.g. iOS]: Reproduced on both Fedora and MacOS
 - Browser [e.g. chrome, safari]. Jest
 - Version [e.g. 22]: 27.5.1

**Comment by ericfrank-stripe** [comment]
TL;DR: The correct fix with the present toolkit is adding the `await update();` call at the end of the test as you've described.

This issue is happening not because the state of `MyComponent` is carrying over to the `MyComponent` in the second test. It's happening because the version of `MyComponent` in the first test is still executing when the second test starts, which interrupts and prolongs the second test so that it has time for the `useEffect` to run and the `2` to show up. This can be demonstrated by using two different components instead of one:

```
export const MyComponent = () => {
  // useState starts with 1
  const [val, setval] = useState(1);

  console.log("MyComponent", { val });

  // And after useEffect it is 2
  useEffect(() => {
    setval(2);
  });
  return <div>{val}</div>
};

export const MyOtherComponent = () => {
  // useState starts with 1
  const [val, setval] = useState(1);

  console.log("MyOtherComponent", { val });

  // And after useEffect it is 2
  useEffect(() => {
    setval(2);
  });
  return <div>{val}</div>
};


describe('Reproduction', () => {
    it('identical 1', async () => {
        console.log("Test 1 begin");
        const thing = render(<MyComponent/>);
        const {wrapper} = thing;
        expect(wrapper).toContainText("1")
    });

    it('identical 2', async () => {
        console.log("Test 2 begin");
        const thing = render(<MyOtherComponent/>);
        const {wrapper} = thing;
        expect(wrapper).toContainText("1")
    });
});
```

Note the console logs I added. Running this test fails the second test with the following log output:
```
Test 1 begin
MyComponent { val: 1 } // We expect 1 and exit the test successfully, but the component is still running a useEffect
Test 2 begin
MyOtherComponent { val: 1 } // The second component renders a 1 and should succeed here, but...
MyComponent { val: 2 } // The component from the first test is still running and blocking the flow of the second test
MyOtherComponent { val: 2 } // And now the second test has had time to render a 2, so it fails
MyComponent { val: 2 }
MyOtherComponent { val: 2 }
```

This is a very common issue in all Jest tests because each tests executes in a common environment (and JS is single threaded). After each test, you should wait for the React tree to finish executing before moving on the next step. It may be possible for us to write a helper that universally stops the execution of each test (we've looked into it), but for now, using the `await update()` helper is the best solution.
